[Cocoa] [GPUP] Compressed fonts are decompressed twice, which is one too many times

litherum opened the pull request:
#### 21ab6bc788f21a099b61fa1ead9f7ba048d7c687
<pre>
[Cocoa] [GPUP] Compressed fonts are decompressed twice, which is one too many times
<a href="https://bugs.webkit.org/show_bug.cgi?id=257401">https://bugs.webkit.org/show_bug.cgi?id=257401</a>
rdar://109910582

Reviewed by David Kilzer.

Previously, the way we were sharing web fonts between the web process and the GPU process is
by simply transferring the bytes of the file, and asking Core Text to open the data. Both
processes need to use the font, so this means Core Text was being asked to open the data twice.
This actually has somewhat significant cost, as many fonts are compressed, and simply opening
it has to decompress the file.

This patch uses an SPI function to produce the decompressed bytes, without opening the font.
This allows us to retain the decompressed bytes, and send them to the GPU process. By
decompressing early, we can decompress once, and then use the decompressed bytes everywhere.
This means we only have to decompress fonts once.

* Source/WebCore/PAL/pal/spi/cf/CoreTextSPI.h:
* Source/WebCore/platform/graphics/coretext/FontCustomPlatformDataCoreText.cpp:
(WebCore::createFontCustomPlatformData):

Canonical link: <a href="https://commits.webkit.org/264724@main">https://commits.webkit.org/264724@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/af065d4519de344594e1112b853b7baec6ed94c8

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/8455 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/8748 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/8967 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/10122 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/8502 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/8463 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/10738 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/8678 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/11368 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/8601 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/9643 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/7633 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/10276 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/6941 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/7736 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/15286 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/8062 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/7882 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/11233 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/8354 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/6816 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/7641 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2043 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/11851 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/8099 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

ddkilzer commented on Source/WebCore/platform/graphics/coretext/FontCustomPlatformDataCoreText.cpp:
```diff
@@ -74,19 +74,33 @@ RefPtr<FontCustomPlatformData> createFontCustomPlatformData(SharedBuffer& buffer
-                auto candidate = static_cast<CTFontDescriptorRef>(CFArrayGetValueAtIndex(array.get(), i));
-                auto postScriptName = adoptCF(static_cast<CFStringRef>(CTFontDescriptorCopyAttribute(candidate, kCTFontNameAttribute)));
+                auto candidate = static_cast<FPFontRef>(CFArrayGetValueAtIndex(array.get(), i));
+                auto postScriptName = adoptCF(static_cast<CFStringRef>(FPFontCopyPostScriptName(candidate)));
```
Why is the `static_cast<CFStringRef>()` needed here when `FPFontCopyPostScriptName()` is declared to return `CFStringRef`?

ddkilzer reviewed (approved):
r=me, but consider removing unneeded `static_cast<CFStringRef>()`.

mattwoodrow commented on Source/WebCore/platform/graphics/coretext/FontCustomPlatformDataCoreText.cpp:
```diff
@@ -74,19 +74,33 @@ RefPtr<FontCustomPlatformData> createFontCustomPlatformData(SharedBuffer& buffer
+
+    RefPtr<SharedBuffer> protectedBuffer = &buffer;
+    RetainPtr<CTFontDescriptorRef> fontDescriptor;
+    if (font) {
```
`font` can't be null at this point, right?

litherum committed:
[Cocoa] [GPUP] Compressed fonts are decompressed twice, which is one too many times
https://bugs.webkit.org/show_bug.cgi?id=257401
rdar://109910582

Reviewed by David Kilzer.

Previously, the way we were sharing web fonts between the web process and the GPU process is
by simply transferring the bytes of the file, and asking Core Text to open the data. Both
processes need to use the font, so this means Core Text was being asked to open the data twice.
This actually has somewhat significant cost, as many fonts are compressed, and simply opening
it has to decompress the file.

This patch uses an SPI function to produce the decompressed bytes, without opening the font.
This allows us to retain the decompressed bytes, and send them to the GPU process. By
decompressing early, we can decompress once, and then use the decompressed bytes everywhere.
This means we only have to decompress fonts once.

* Source/WebCore/PAL/pal/spi/cf/CoreTextSPI.h:
* Source/WebCore/platform/graphics/coretext/FontCustomPlatformDataCoreText.cpp:
(WebCore::createFontCustomPlatformData):

Canonical link: https://commits.webkit.org/264724@main
